GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -458,37 +458,98 @@ static void StraightenLanes(int startIndex, int lastIndex, IList<RevisionGraphRo
+                        }
+
+                        int straightenedCurrentLane = currentLane + 1;
+                        int lookaheadLane = currentLane;
+                        int nextIndex = currentIndex + 1;
+                        var segmentOrAncestor = localOrderedRowCache[currentIndex].FirstParentOrSelf(revisionGraphSegment);
+
+                        for (int lookaheadIndex = nextIndex; lookaheadLane == currentLane && lookaheadIndex <= Math.Min(currentIndex + _straightenLanesLookAhead, lastIndex); ++lookaheadIndex)
                         {
-                            int straightenedCurrentLane = currentLane + 1;
-                            int lookaheadLane = currentLane;
-                            int nextIndex = currentIndex + 1;
-                            for (int lookaheadIndex = nextIndex; lookaheadLane == currentLane && lookaheadIndex <= Math.Min(currentIndex + _straightenLanesLookAhead, lastIndex); ++lookaheadIndex)
+                            RevisionGraphRow lookaheadRow = localOrderedRowCache[lookaheadIndex];
+                            lookaheadLane = lookaheadRow.GetLaneIndexForSegment(segmentOrAncestor);
+                            if ((lookaheadLane == straightenedCurrentLane) || (lookaheadLane > straightenedCurrentLane && previousLane == straightenedCurrentLane))
                             {
-                                lookaheadLane = localOrderedRowCache[lookaheadIndex].GetLaneIndexForSegment(revisionGraphSegment);
-                                if ((lookaheadLane == straightenedCurrentLane) || (lookaheadLane > straightenedCurrentLane && previousLane == straightenedCurrentLane))
+                                if (MovingCausesWidthIncrease(currentIndex, lookaheadIndex, currentLane))
                                 {
-                                    currentRow.MoveLanesRight(currentLane);
-                                    for (; nextIndex < lookaheadIndex; ++nextIndex)
-                                    {
-                                        localOrderedRowCache[nextIndex].MoveLanesRight(currentLane);
-                                    }
+                                    break;
+                                }
 
-                                    moved = true;
+                                if (MovingIncreasesASegmentBend(currentRow, previousRow, currentLane))
+                                {
                                     break;
                                 }
+
+                                for (int index = currentIndex; index < lookaheadIndex; ++index)
+                                {
+                                    localOrderedRowCache[index].MoveLanesRight(currentLane);
+                                }
+
+                                moved = true;
+                                break;
                             }
+
+                            segmentOrAncestor = lookaheadRow.FirstParentOrSelf(segmentOrAncestor);
```

Comment:
```suggestion
                            segmentOrAncestor = lookaheadRow.FirstParentOrSelf(segmentOrAncestor);
                        }
                        
                        if (moved)
                        {
                            break;
```
This was missing in the original implementation.

Reply:
Done